### DISCUSSION TITLE
Single User Connections

### DISCUSSION THREAD

**Discussion by tmayoff** [OpenDDS/OpenDDS#3856] [Q&A] [answered]
I'm not sure how to phrase this, but before I go implementing something I wanted to know if there's an existing way or common approach to this. We'd like the ability for a user/operator of our system to 'connect' to a single machine. Essentially filtering out all DDS samples except those coming from the user, and the user filter everything but samples from the machine. We still want the user and remote machine to utilise DDS and it's PUB/SUB pattern. My first thought was to add an identifier to all the topics and filter all of the subscriber, and update those filters as the 'connected' user changes (but this poses an issue with how to update filters with multiple parameters). Or I also thought it could be done via domainIDs  but it seems like that might be abusing the concept of domains. I also briefly looked into DDSSecurity but am not sure if that'll accomplish what I'm after. Any advice or help would be appreciated

**Comment by simpsont-oci** [comment, marked as answer]
How important is reducing traffic in your situation?

The easiest solution, from the perspective of DDS, is probably creating ContentFilteredTopics and having the machine filter by user and the user filter by machine. Currently, this doesn't really reduce overall traffic, so samples will still be sent everywhere and then filtered on the receiving side. But perhaps that's acceptable for your use case.

Another potential solution would be to use one topic to establish the "connection" as you call it by agreeing on one ore more new topic names for that connection (where presumably the types are known beforehand). Then the machine and the user could both create readers and writers using those new topic names and, as long as the topic names are unique, the traffic would be limited to those participants (machine / user) using the new topics.

Security will be helpful if you need authentication, access control, or encryption, but so far it just seems like you're more interested in filtering for the sake of cleaner code / reduced traffic rather than security concerns.

**Reply by tmayoff** [reply]
Thanks for the reply! 

I thought that with Content Filtered Topics the filtering is done on the publisher side (at least that's what is says in the Dev Guide) so wouldn't that reduce overall traffic?

**Reply by mitza-oci** [reply]
Content Filtering may happen on the publishing side (unless configuration says not to), based on the kinds of transports in use and how the writer can reach any interested (not filtered-out) readers using those transports.

**Reply by tmayoff** [reply]
Rereading the QoS policies for other things I noticed the Partition QoS, would that also work as an alternative to at least the multiple topic names, giving each 'connection' a unique ID/key and place that in the partition QoS for the DataReader/Writer. Also during discovery presumably any other reader could just see the generated topic name (or partition name) so it wouldn't be 'secure' would DDS Security help with this?

Also I wanted to clarify the difference between the datawriter and a publisher. When should we create a new publisher vs a new datawriter. Since the publisher doesn't contain any info on the topic, all of that is in the datawriter should we just have many writers and a couple publisher only creating a new publisher if we need a different QoS policy (or a different listener)? (I also assume the same logic would apply to the subscriber/datareader side)

**Reply by simpsont-oci** [reply]
Interestingly, and partially in answer to your second question, Partition QoS can only be applied at the publisher and subscriber level, not at the datawriter or datareader level. So that's one potential reason why you might want to create more than one publisher or subscriber per participant. Otherwise, if you're not interested in other publisher and subscriber-level QoS policies (e.g presentation QoS) then you can usually get by with a single publisher or subscriber per participant.

DDS Security has options which allow for encrypting discovery traffic, so yes, it could help with hiding partition or topic names, etc.
